doc/developer-guides/sw_design_guidelines.rst

```diff
@@ -0,0 +1,494 @@
+   | programming           |               |          | Plan to do range check in hypercalls |
+   |                       |               |          | and HW capability checks.            |
+   +-----------------------+---------------+----------+--------------------------------------+
+
+
+Error Handling Methods
+----------------------
+
+The error handling methods used in ACRN hypervisor on architecture level are
+shown below.
+
+**Invoke default fatal error handler**
+  The hypervisor shall invoke default fatal error handler when the below cases
+  occur. Customers can define platform-specific handlers, allowing them to
+  implement additional error reporting (mostly to hardware) if required. The
+  default fatal error handler will invoke platform-specific handlers defined by
+  users at first, then it will panic the system.
+
+  This method applies to the following cases:
+
+  - Related hardware resources are unavailable.
+  - Boot information is invalid during platform initialization.
+  - Unexpected exception occurs in root mode due to hardware failures.
+  - Failures occur in the VM dedicated for error handling.
+
+**Return error code**
+  The hypervisor shall return error code to the VM when the below cases occur.
+  The error code shall indicate the error type detected (e.g. invalid parameter,
+  device not found, device busy, resource unavailable, etc).
+
+  This method applies to the following case:
+
+  - The hypercall parameter from the VM is invalid.
+
+**Inform the safety VM through specific register or memory area**
+  The hypervisor shall inform the safety VM through specific register or memory
+  area when the below cases occur. The VM will decide how to handle the related
+  error. This shall only be done after the VM (Safety OS or Service OS)
+  dedicated for error handling have started.
+
+  This method applies to the following cases:
+
+  - Machine check errors occur due to hardware failures.
+
+  - Unexpected VM entry failures occur, where the VM is not the one dedicated
+    for error handling.
+
+**Crash the system via ASSERT**
+  The hypervisor can crash the system when the below cases occur. It shall
+  only be used for debug and used to check pre-conditions and post-conditions
+  to catch design errors.
+
+  This method applies to the following case:
+
+  - Software design errors occur.
+
+
+Rules of Error Detection and Error Handling
+-------------------------------------------
+
+The rules of error detection and error handling on architecture level is shown
+in :numref:`rules_arch_level` below.
+
+.. table:: Rules of Error Detection and Error Handling on Architecture Level
+   :align: center
+   :widths: auto
+   :name: rules_arch_level
+
+   +--------------------+-------------------------+--------------+---------------------------+-------------------------+
+   | **Resource Class** | **Failure Mode**        | **Error      | **Error Handling Policy** | **Example**             |
```

Comment:
The table heading is **bold** by default, so you can leave off the bold indicators.